File: apps/router/app/tracing.py
from __future__ import annotations

import logging
from typing import Any

from opentelemetry import context, propagate, trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import (
    BatchSpanProcessor,
    ConsoleSpanExporter,
    SimpleSpanProcessor,
)
from opentelemetry.trace import Span, Status, StatusCode, Tracer

from app.config import Settings
from app.config import settings as default_settings

logger = logging.getLogger(__name__)

_initialized = False


def init_tracing(settings: Settings | None = None) -> None:
    """Configure tracer provider once. No OTLP exporter unless endpoint is set."""
    global _initialized
    if _initialized:
        return

    cfg = settings or default_settings
    if not cfg.otel_enabled:
        _initialized = True
        return

    resource = Resource.create({"service.name": cfg.otel_service_name})
    provider = TracerProvider(resource=resource)

    endpoint = (cfg.otel_exporter_otlp_endpoint or "").strip().rstrip("/")
    if endpoint:
        # OTLP HTTP exporter expects the full traces URL.
        exporter = OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces")
        provider.add_span_processor(BatchSpanProcessor(exporter))
        logger.info("otel otlp exporter endpoint=%s", endpoint)
    if cfg.otel_console_exporter:
        provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))

    trace.set_tracer_provider(provider)
    _initialized = True


def get_tracer(name: str = "router") -> Tracer:
    return trace.get_tracer(name)


def extract_context(headers: Any) -> context.Context:
    """Extract W3C trace context from inbound HTTP headers."""
    carrier: dict[str, str] = {}
    try:
        for key, value in headers.items():
            if isinstance(value, str):
                carrier[str(key).lower()] = value
            elif isinstance(value, (list, tuple)) and value:
                carrier[str(key).lower()] = str(value[0])
    except Exception:  # noqa: BLE001
        return context.get_current()
    return propagate.extract(carrier)


def attach_context(ctx: context.Context) -> Any:
    return context.attach(ctx)


def detach_context(token: Any) -> None:
    try:
        context.detach(token)
    except Exception:  # noqa: BLE001
        pass


def set_span_error(span: Span, err: BaseException | str) -> None:
    if isinstance(err, BaseException):
        span.record_exception(err)
        span.set_status(Status(StatusCode.ERROR, str(err)))
    else:
        span.set_status(Status(StatusCode.ERROR, err))


def set_span_ok(span: Span) -> None:
    span.set_status(Status(StatusCode.OK))

File: apps/gateway/src/tracing.ts
import { context, propagation, SpanStatusCode, trace, type Span, type Tracer } from "@opentelemetry/api";
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks";
import { W3CTraceContextPropagator } from "@opentelemetry/core";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { Resource } from "@opentelemetry/resources";
import {
  BasicTracerProvider,
  BatchSpanProcessor,
  ConsoleSpanExporter,
  SimpleSpanProcessor,
} from "@opentelemetry/sdk-trace-base";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";

export type TracingConfig = {
  enabled: boolean;
  serviceName: string;
  /** OTLP HTTP base, e.g. http://jaeger:4318 — exporter posts to /v1/traces */
  otlpEndpoint: string;
  /** Also print spans to stdout (local debug). */
  consoleExporter: boolean;
};

let provider: BasicTracerProvider | null = null;
let contextManagerReady = false;

function ensureContextManager(): void {
  if (contextManagerReady) return;
  context.setGlobalContextManager(new AsyncLocalStorageContextManager().enable());
  contextManagerReady = true;
}

export function loadTracingConfig(env: NodeJS.ProcessEnv = process.env): TracingConfig {
  const enabledRaw = (env.OTEL_ENABLED ?? "true").trim().toLowerCase();
  const enabled = !["0", "false", "no", "off"].includes(enabledRaw);
  const consoleRaw = (env.OTEL_CONSOLE_EXPORTER ?? "false").trim().toLowerCase();
  return {
    enabled,
    serviceName: (env.OTEL_SERVICE_NAME ?? "gateway").trim() || "gateway",
    otlpEndpoint: (env.OTEL_EXPORTER_OTLP_ENDPOINT ?? "").trim().replace(/\/$/, ""),
    consoleExporter: ["1", "true", "yes", "on"].includes(consoleRaw),
  };
}

/** Initialize the tracer provider once (safe to call repeatedly). */
export function initTracing(config: TracingConfig = loadTracingConfig()): void {
  ensureContextManager();
  propagation.setGlobalPropagator(new W3CTraceContextPropagator());
  if (!config.enabled) {
    return;
  }
  if (provider) {
    return;
  }

  const resource = new Resource({
    [ATTR_SERVICE_NAME]: config.serviceName,
  });
  const tracerProvider = new BasicTracerProvider({ resource });

  if (config.otlpEndpoint) {
    const url = `${config.otlpEndpoint}/v1/traces`;
    tracerProvider.addSpanProcessor(
      new BatchSpanProcessor(new OTLPTraceExporter({ url })),
    );
  }
  if (config.consoleExporter) {
    tracerProvider.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));
  }

  trace.setGlobalTracerProvider(tracerProvider);
  provider = tracerProvider;
}

export async function shutdownTracing(): Promise<void> {
  if (!provider) return;
  await provider.shutdown();
  provider = null;
}

export function getTracer(name = "gateway"): Tracer {
  return trace.getTracer(name);
}

/** Inject W3C trace headers. Pass `span` when not running under an active context. */
export function injectTraceHeaders(headers: Record<string, string>, span?: Span): void {
  const ctx = span ? trace.setSpan(context.active(), span) : context.active();
  propagation.inject(ctx, headers, {
    set: (carrier, key, value) => {
      carrier[key] = value;
    },
  });
}

export function startRequestSpan(
  tracer: Tracer,
  name: string,
  attributes: Record<string, string | number | boolean>,
): Span {
  return tracer.startSpan(name, { attributes });
}

export function endSpanOk(span: Span): void {
  span.setStatus({ code: SpanStatusCode.OK });
  span.end();
}

export function endSpanError(span: Span, err: unknown): void {
  if (err instanceof Error) {
    span.recordException(err);
  }
  span.setStatus({
    code: SpanStatusCode.ERROR,
    message: err instanceof Error ? err.message : String(err),
  });
  span.end();
}

export function withSpanContext<T>(span: Span, fn: () => T): T {
  return context.with(trace.setSpan(context.active(), span), fn);
}
